src/Client/ClientBase.h

```diff
@@ -0,0 +1,204 @@
+#pragma once
+
+#include <boost/program_options.hpp>
+#include <Poco/Util/Application.h>
+#include <Interpreters/Context.h>
+#include <Common/ProgressIndication.h>
+#include <Client/Suggest.h>
+#include <Client/QueryFuzzer.h>
+#include <Common/ShellCommand.h>
+
+namespace po = boost::program_options;
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int NOT_IMPLEMENTED;
+}
+}
+
+
+namespace DB
+{
+
+class ClientBase : public Poco::Util::Application
+{
+
+public:
+    using Arguments = std::vector<String>;
+
+    void init(int argc, char ** argv);
+
+    int main(const std::vector<String> & /*args*/) override;
+
+protected:
+    /*
+     * Run interactive or non-interactive mode. Depends on:
+     *  - processSingleQuery
+     *  - processMultiQuery
+     *  - processWithFuzzing
+     */
+    void runNonInteractive();
+
+    /// Pass lambda, which executes passed function and catches/processes exceptions, returns result of passed function.
+    void runInteractive(std::function<bool(std::function<bool()>)> try_process_query_text);
```

Comment:
Callback here is needed only because real client need to reconnect on error, right?

I was thinking about adding an interface for Connection, and create LocalConnection (or similar) that will be used only in conjunction with LocalServer, or ClientBase can be templated by the Connection (to avoid virtual overhead regular Connections), this way callback can be dropped.

But this may be overkill for this task I guess...

Reply:
> Callback here is needed only because real client need to reconnect on error, right?

yep

>  and create LocalConnection (or similar) that will be used only in conjunction with LocalServer

You mean LocalConnection to be a something like a wrapper for calling executeQuery? It is not intuitive that there should be such class to be used with clickhouse-local. May be I just did not get the idea...

Reply:
>You mean LocalConnection to be a something like a wrapper for calling executeQuery?

Yep.

>It is not intuitive that there should be such class to be used with clickhouse-local.

OK, I'm not sure about this idea too, so let's ignore it.

>May be I just did not get the idea...

The purpose of this was to use generic code instead of hooking it with some callbacks.